### PR TITLE
Stokhos: Make sure deep_copy uses contiguous layouts where intended

### PR DESCRIPTION
**Pull request by masterleinad**
@trilinos/stokhos @etphipp

## Motivation
https://github.com/kokkos/kokkos/pull/5209 added a check for the layout to be used in Kokkos::View constructors to be constructible from just extents. The only standard layout for which this property doesn't hold is `LayoutStride`. The cases fixed below mean to create a contiguous representation of a given `View` in case the View is not contiguous in memory but the newly created Views just copy the layout from the old ones which makes the new Views also non-contiguous in memory in case of `LayoutStride`. In the end, the new `View` is initialized only from extents which again doesn't do the right thing for `LayoutStride` since its constructor expects extents and strides alternating.
The proposed fix is to just take the preferred layout of the associated execution space which is either `LayoutLeft` or `LayoutRight` and in particular constructible from extents only.

## Related Issues

Fixes https://github.com/kokkos/kokkos/issues/5243.

## Testing
Compiling and running all Stokhos tests with the `Serial` backend and `Kokkos` `develop`.

### CODE REVIEW

**Review by etphipp — APPROVED** [review]

**Comment by ndellingwood** [comment]
@trilinos/framework the Trilinos_PR_intel-17.0.1 failing build#485 https://trilinos-cdash.sandia.gov/build/302094/configure was due to configure issue unrelated to this PR, were any changes made for the intel/17 jobs/env or is this a spurious glitch that just requires retest? Here is the error:
```
...
-- CMAKE_CXX_COMPILER_ID='Intel'
-- CMAKE_CXX_COMPILER_VERSION='17.0.1.20161005'
-- The Fortran compiler identification is Intel 17.0.1.20161005
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - failed
-- Check for working Fortran compiler: /projects/sems/install/rhel7-x86_64/sems/compiler/intel/17.0.1/mpich/3.2/bin/mpif90
-- Check for working Fortran compiler: /projects/sems/install/rhel7-x86_64/sems/compiler/intel/17.0.1/mpich/3.2/bin/mpif90 - broken
CMake Error at /projects/sems/install/rhel7-x86_64/sems/utility/cmake/3.19.1/share/cmake-3.19/Modules/CMakeTestFortranCompiler.cmake:51 (message):
  The Fortran compiler

    "/projects/sems/install/rhel7-x86_64/sems/compiler/intel/17.0.1/mpich/3.2/bin/mpif90"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /scratch/trilinos/jenkins/ascic142/workspace/Trilinos_PR_intel-17.0.1/pull_request_test/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/projects/sems/install/rhel7-x86_64/sems/utility/ninja_fortran/1.10.0/bin/ninja cmTC_40e3b && [1/4] Building Fortran preprocessed CMakeFiles/cmTC_40e3b.dir/testFortranCompiler.f-pp.f
    [2/4] Generating Fortran dyndep file CMakeFiles/cmTC_40e3b.dir/Fortran.dd
    [3/4] Building Fortran object CMakeFiles/cmTC_40e3b.dir/testFortranCompiler.f.o
    FAILED: CMakeFiles/cmTC_40e3b.dir/testFortranCompiler.f.o 
    /projects/sems/install/rhel7-x86_64/sems/compiler/intel/17.0.1/mpich/3.2/bin/mpif90  -I. -fPIE -c CMakeFiles/cmTC_40e3b.dir/testFortranCompiler.f-pp.f -o CMakeFiles/cmTC_40e3b.dir/testFortranCompiler.f.o
    ifort: warning #10352: The directory '/tmp' is full.  Please check disk space.
    ifort: error #10001: could not find directory in which g++ resides
    ninja: build stopped: subcommand failed.
    
```

**Comment by ndellingwood** [comment]
The gcc/7.2.0 and clang/10.0 failures are likely related to c++17 changes to those builds and hopefully resolved by retest after merge of PRs #10786 and #10780

**Comment by ndellingwood** [comment]
@trilinos/framework the failing `TpetraTSQR_SequentialTsqr_contiguousCacheBlocks_MPI_1` in the cuda/11.4.2 non-uvm build (great to see the new build!) should be completely unrelated to the changes in this PR - this PR modified two lines in Stokhos and should have no impact on Tpetra TSQR (a grep for Stokhos through `packages/tpetra/tsqr` shows no Stokhos usage or inclusion of headers in the src nor tests). 

Failure output from cdash https://trilinos-cdash.sandia.gov/test/8047617 for reference:
```
NodeTsqr verify/benchmark test options:
  - NodeTsqr: SequentialTsqr
  - numRows: 100000
  - numCols: 10
  - numTrials: 10
  - testReal: true
  - testComplex: false
  - cacheSizeHint: 5000
  - contiguousCacheBlocks: true
  - printFieldNames: true
  - printTrilinosTestStuff: true
  - humanReadable: false
  - verbose: false
  - saveMatrices: false
%method,scalarType,numRows,numCols,cacheSizeHint,contiguousCacheBlocks,frobA,absFrobResid,absFrobOrthog
CUSOLVER,float,100000,10,0,false,1.15429,4.47428e-07,6.45184e-07
LAPACK,float,100000,10,0,false,1.15429,2.71137e-07,7.84201e-07
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaMemcpyAsync(dst, src, n, cudaMemcpyDefault, instance.cuda_stream()) error( cudaErrorIllegalInstruction): an illegal instruction was encountered /gpfs/trilinos/workspace/Trilinos_PR_cuda-11.4.2-uvm-off/Trilinos/packages/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:99
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec noticed that process rank 0 with PID 0 on node ascicgpu051 exited on signal 6 (Aborted).
--------------------------------------------------------------------------
```

Is this possibly a preexisting failure that slipped through with testing with uvm enabled?

**Comment by masterleinad** [comment]
Regarding https://github.com/trilinos/Trilinos/pull/10783#issuecomment-1191708482
the problem is https://github.com/trilinos/Trilinos/blob/7919a38aa0454197eb3889ca184950c4d25d949a/packages/stokhos/src/sacado/kokkos/pce/KokkosExp_View_UQ_PCE_Contiguous.hpp#L1549-L1574

In this particular case, this pull request changes the layout from `LayoutLeft` to `LayoutRight` and we are copying in the end between a `LayoutLeft` destination view and a `LayoutRight` source view. They end up not having the same extents since the View assignment puts the `sacado_size` first for `LayoutLeft` and last otherwise (here `LayoutRight`). Supposedly this assignment should also work for `LayoutStride` and do the same thing as `LayoutRight` but I'm not sure if this actually works correctly. Note that the original issue was that `deep_copy` was instantiated with a `LayoutStride`, though. Maybe, someone from the Stokhos/Sacado team can clarify how this is supposed to work.

**Comment by masterleinad** [comment]
Based on my observations in https://github.com/trilinos/Trilinos/pull/10783#issuecomment-1192641737, I opted for replacing `LayoutStride` with `LayoutRight` instead of using the execution space's default layout.

**Comment by ndellingwood** [comment]
Thanks for the update @masterleinad !

@etphipp when you get a chance can you re-review the updates to the PR (changes dismissed the previous approval)?

**Review by etphipp — APPROVED** [review]
Sorry for the delay.  I was on vacation.

**Pull request merged**
